[IMP] payment providers: PayPal (#4396)
odoo/documentation: content/applications/finance/payment_acquirers/paypal.rst

======
PayPal
======

`Paypal <https://www.paypal.com/>`_ is an American online payments acquirer available worldwide. As
it does not charge any subscription fee and creating an account is easy, it is recommended for Odoo
starters.

> Settings in Odoo

Author: Thought that's what you meant in one of the previous comments, I also found it strange

> Settings in Odoo

Reviewer: I anything I'd have said it was a good idea^^

- :guilabel:`Use IPN`: enable, as it is :ref:`required <paypal/auto-return>` for PayPal to work
  properly in Odoo.

To set the :guilabel:`PDT Identity Token`, switch to :ref:`developer mode <developer-mode>` and
retrieve the token by following the configuration steps at :ref:`paypal/pdt`.

.. note::
   The PayPal **Merchant ID** is not required in Odoo.
.. important::
   If you are trying PayPal as a test, using a :ref:`PayPal Sandbox account <paypal/testing>`,
   change the :guilabel:`State` to :guilabel:`Test Mode`. We recommend doing this on a test Odoo
   database rather than on your main database.

Extra fees
----------

You can charge **extra fees** to customers choosing to pay with PayPal in order to cover the
transaction fees PayPal charges you. Once redirected to Paypal, your customer sees an extra amount
applied to the order amount.

To activate this, go to the :guilabel:`Fees` tab in Odoo and activate :guilabel:`Add Extra Fees`.

.. note::
   -  You can refer to `Paypal Fees <https://www.paypal.com/webapps/mpp/paypal-fees>`_ to set up
      fees.
   - `Traders in the EU <https://europa.eu/youreurope/citizens/consumers/shopping/pricing-payments/index_en.htm>`_

>       are not allowed to charge extra fees for paying with credit cards.

Reviewer: This seems to be cutting the sentence in half.
```suggestion
     are not allowed to charge extra fees for paying with credit cards.
```

.. _paypal/testing:

Settings in PayPal
==================

To access your PayPal account settings, log into PayPal, open the :guilabel:`Account Settings`, and
open the :guilabel:`Website payments` menu.

.. image:: paypal/paypal-account.png
   :align: center
   :alt: PayPal account menu

.. _paypal/auto-return:

Auto Return
-----------

The **Auto Return** feature automatically redirects customers to Odoo once the payment is processed.

From :guilabel:`Website payments`, go to :menuselection:`Website preferences --> Update --> Auto
return for website payments --> Auto return` and select :guilabel:`On`. Enter the address of your
Odoo database (e.g., `https://yourcompany.odoo.com`) in the :guilabel:`Return URL` field, and
:guilabel:`Save`.

.. note::
   Any URL will do the job. Odoo only needs the setting to be enabled since it uses another URL.
.. important::
   Note that for PayPal to work **in Odoo**, the options :ref:`Auto Return <paypal/auto-return>`,

>    :ref:`PDT <paypal/pdt>`, and :ref:`IPN <paypal/ipn>` **must** all be enabled.

Reviewer: This could even be moved above "Auto Return".

======
PayPal
======

`Paypal <https://www.paypal.com/>`_ is an American online payment acquirer available worldwide, and

> one of the that does not charge a subscription fee.

Author: Fixed

======
PayPal
======

`Paypal <https://www.paypal.com/>`_ is an American online payments acquirer available worldwide. As
it does not charge any subscription fee and creating an account is easy, it is recommended for Odoo
starters.

Settings in PayPal
==================

To access your PayPal account settings, log into PayPal, open the :guilabel:`Account Settings`, and
open the :guilabel:`Website payments` menu.

.. image:: paypal/paypal-account.png
   :align: center
   :alt: PayPal account menu

.. important::
   Note that for PayPal to work **in Odoo**, the options :ref:`Auto Return <paypal/auto-return>`,

>    :ref:`PDT <paypal/pdt>`, and :ref:`IPN <paypal/ipn>` **must** all be enabled.

Reviewer 2: then maybe the pragraph could be re-written as I don't understand why  you added the info that should be performed from the settings it in an important note. We can discuss it if you want :) 
I would remove the important note and explain what should be done from the Paypal settings:
"To access your PayPal account settings, log into PayPal, open the Account Settings, and open the Website payments menu. From here, you must the options XXX must be enabled."

>    :ref:`PDT <paypal/pdt>`, and :ref:`IPN <paypal/ipn>` **must** all be enabled.

Author: Because those options in PayPal are not mandatory for PayPal itself to work, sometimes it's preferences. Customers who used PayPal in other instances or with other software may think some options can be left disabled in Odoo too, which is *not* the case.

Some options *have* to be enabled for the PayPal integration to work with Odoo, even if the customer used to do it differently in its previous software.

.. _paypal/ipn:

Instant Payment Notification (IPN)
----------------------------------

:abbr:`IPN (Instant Payment Notifications)` is similar to **PDT**, but allows for more
notifications, such as chargeback notifications. To enable **IPN**, go to :menuselection:`Website
payments --> Instant payment notifications --> Update` and click :guilabel:`Choose IPN settings`.
Enter a :guilabel:`Notification URL`, select :guilabel:`Receive IPN messages (Enabled)`, and
:guilabel:`Save`.

PayPal Account Optional
-----------------------

We advise you not to prompt customers to log in with a PayPal account upon payment. It is better and
more accessible for customers to pay with a debit/credit card. To disable that prompt, go to
:menuselection:`Account Settings --> Website payments --> Update` and select :guilabel:`On` for
:guilabel:`PayPal account optional`.

Payment Messages Format
-----------------------

In case you use accented characters (or anything else than primary Latin characters) for customer
names or addresses, then you **must** configure the encoding format of the payment request sent by

> Odoo to PayPal. Otherwise, some transactions fail without notice.

Author: Both, and only those I believe

starters.

Settings in PayPal
==================

.. _paypal/auto-return:

Auto Return
-----------

The **Auto Return** feature automatically redirects customers to Odoo once the payment is processed.

From :guilabel:`Website payments`, go to :menuselection:`Website preferences --> Update --> Auto
return for website payments --> Auto return` and select :guilabel:`On`. Enter the address of your
Odoo database (e.g., `https://yourcompany.odoo.com`) in the :guilabel:`Return URL` field, and
:guilabel:`Save`.

.. note::
   - Any URL will do the job. Odoo only needs the setting to be enabled since it uses another URL.
   - Turning :guilabel:`Auto Return` *off* disables the :ref:`Payment Data Transfer
     <paypal/pdt>` feature.

.. _paypal/pdt:

Payment Data Transfer (PDT)
---------------------------

:guilabel:`Payment Data Transfer` allows to receive payment confirmations, displays the payment
status to the customers, and verifies the authenticity of the payments. :abbr:`PDT (Payment Data
Transfer)`. From :menuselection:`Website preferences --> Update`, scroll down to :guilabel:`Payment
data transfer` and select :guilabel:`On` (if desired). **PDT** is *not* mandatory, and :ref:`IPN

> (Instant Payment Notifications) <paypal/ipn>` can be used instead.

Reviewer: This is wrong. The field might not be flagged as required in Odoo 15 but it has been fixed in later versions. If PDT is not enabled in any Odoo version, the PayPal integration will *not* work properly. See https://github.com/odoo/odoo/blob/15.0/addons/payment_paypal/controllers/main.py#L87-L95

.. note::
   To use **PDT**, :ref:`Auto Return <paypal/auto-return>` must also be enabled.

> .. tip::
>    PayPal **recommends** using **IPN**. For more information on **PDT** and **IPN**, click `here. <https://developer.paypal.com/api/nvp-soap/ipn/IPNPDTAnAlternativetoIPN/>`_

Reviewer: The recommendation doesn't apply to the API Odoo integrates with and is bad practice anyway for integrators like Odoo databases. We should make it clear that all of Auto return, PDT and IPN **must** be enabled for Odoo to work properly.
```suggestion
```

> PDT Identity Token

Reviewer: Could be removed.

~~~~~~~~~~~~~~~~~~

PayPal displays your **PDT Identity Token** as soon as :ref:`Auto return <paypal/auto-return>` and
:ref:`Payment Data Transfer (PDT) <paypal/pdt>` are enabled. If you need the **PDT Identity Token**,

> disable and re-enable :guilabel:`Payment data transfer` to display the token again.

Reviewer: Sounds like a `.. tip::`.

.. _paypal/ipn:

Instant Payment Notification (IPN)
----------------------------------

:abbr:`IPN (Instant Payment Notifications)` is similar to **PDT**, but allows for more
notifications, such as chargeback notifications. To enable **IPN**, go to :menuselection:`Website
payments --> Instant payment notifications --> Update` and click :guilabel:`Choose IPN settings`.
Enter a :guilabel:`Notification URL`, select :guilabel:`Receive IPN messages (Enabled)`, and

> :guilabel:`Save`. **IPN** is an alternative to :ref:`DPT <paypal/pdt>`, and is *not* mandatory.

Reviewer: ```suggestion
:guilabel:`Save`.
```

> .. tip::
>    It is possible to enable both **PDT** and **IPN** simultaneously. More info `here. <https://developer.paypal.com/api/nvp-soap/ipn/IPNPDTAnAlternativetoIPN/>`_

Reviewer: ```suggestion
```

     `link <https://sandbox.paypal.com/cgi-bin/customerprofileweb?cmd=_profile-language-encoding>`_
     to configure the encoding format in a test environment.

Settings in Odoo
================

.. seealso::
   :ref:`payment_acquirers/add_new`

Credentials
-----------

Odoo needs your **API Credentials** to connect with your PayPal account. To do so, go to
:menuselection:`Accounting --> Configuration --> Payment Acquirers` and :guilabel:`Activate` PayPal.
Then, enter your PayPal account credentials in the :guilabel:`Credentials` tab:

- :guilabel:`Email`: the login email address in Paypal;
- :guilabel:`PDT Identity Token`: the key used to verify the authenticity of transactions;

> - :guilabel:`Use IPN`: enable, as it is :ref:`required <paypal/auto-return>` for PayPal to work

Reviewer: ```suggestion
- :guilabel:`Use IPN`: enable, as it is required for PayPal to work
```
Since the warning has moved.

  properly in Odoo.

To set the :guilabel:`PDT Identity Token`, switch to :ref:`developer mode <developer-mode>` and

> retrieve the token by following the configuration steps at :ref:`paypal/pdt`.

Reviewer: In addition, or alternative, I suggest following what was done in the latest payment provider docs where we recommend saving the generated keys for later: https://www.odoo.com/documentation/master/applications/finance/payment_providers/amazon_payment_services.html#configuration-on-aps-dashboard

     `link <https://sandbox.paypal.com/cgi-bin/customerprofileweb?cmd=_profile-language-encoding>`_
     to configure the encoding format in a test environment.

Settings in Odoo
================

.. seealso::

>    - :ref:`payment_acquirers/add_new`

Reviewer: ```suggestion
   :ref:`payment_acquirers/add_new`
```

>    - :ref:`payment_acquirers/add_new`

Reviewer: ```suggestion
   :ref:`payment_acquirers/add_new`
```

> Credentials tab

Reviewer: Please don't structure the page following the tabs in Odoo; it's terrible. Instead, the parent "Payment acquirers" page should rather be purged of this structure too.

---------------

Odoo needs your **API Credentials** to connect with your PayPal account. To do so, go to
:menuselection:`Accounting --> Configuration --> Payment Acquirers` and :guilabel:`Activate` PayPal.
Then, enter your PayPal account credentials in the :guilabel:`Credentials` tab:

- :guilabel:`Email`: the login email address in Paypal;
- :guilabel:`PDT Identity Token`: the key used to verify the authenticity of transactions;

> - :guilabel:`Use IPN`: whether you want to use **Instant Payment Notifications**.

Reviewer: :warning: To drop in the forward-port to saas-16.2.

To set the :guilabel:`PDT Identity Token`, switch to :ref:`developer mode <developer-mode>` and

> retrieve the token by following the configuration steps at :ref:`paypal/pdt`.

Reviewer: Same, 16.2

.. note::

>    The PayPal **Merchant ID** is not required.

Reviewer: Same for 16.2; it won't exist anymore.

.. important::
   If you are trying PayPal as a test, using a :ref:`PayPal Sandbox account <paypal/testing>`,
   change the :guilabel:`State` to :guilabel:`Test Mode`. We recommend doing this on a test Odoo
   database rather than on your main database.

> Fees tab

Reviewer: Same

--------

You can charge **extra fees** to customers choosing to pay with PayPal in order to cover the
transaction fees PayPal charges you. Once redirected to Paypal, your customer sees an extra amount
applied to the order amount.

> To activate this, go to the :guilabel:`Fees` tab in Odoo and activate :guilabel:`Add Extra Fees`.

Reviewer: Note: this should be replaced by a ref to [the new "Extra Fees" section in the parent page](https://www.odoo.com/documentation/16.0/applications/finance/payment_providers.html#extra-fees) starting from 16.0.

Configuration
-------------

> Thanks to PayPal sandbox accounts, you can test the entire payment flow in Odoo.

Reviewer 2: ```suggestion
Thanks to :ref:`PayPal Sandbox account <paypal/testing>`, you can test the entire payment flow in Odoo.
```
